File: web/manga-reader/src/entities/chapter/model/admin/chapterStatus.ts
import type { ChapterStatus } from './chapterAdmin.types';

/**
 * Máquina de estados do capítulo.
 *
 * - `draft`       → edição livre; ponto de partida e de retorno.
 * - `processing`  → páginas em processamento (pipeline de imagens).
 * - `scheduled`   → publicação agendada (exige data futura).
 * - `published`   → único status visível publicamente.
 * - `hidden`      → retirado do site, acessível só no painel admin.
 * - `unavailable` → indisponível (ex.: problema de licenciamento).
 * - `archived`    → arquivado; nunca aparece publicamente.
 */
export const CHAPTER_STATUS_TRANSITIONS: Record<ChapterStatus, ChapterStatus[]> = {
    draft: ['processing', 'scheduled', 'published', 'archived'],
    processing: ['draft', 'published'],
    scheduled: ['draft', 'published', 'archived'],
    published: ['hidden', 'unavailable', 'archived'],
    hidden: ['published', 'archived'],
    unavailable: ['published', 'archived'],
    archived: ['draft'],
};

export const canTransition = (from: ChapterStatus, to: ChapterStatus): boolean => CHAPTER_STATUS_TRANSITIONS[from].includes(to);

/** Apenas `published` aparece no site; todo o resto é restrito ao painel. */
export const isChapterPubliclyVisible = (status: ChapterStatus): boolean => status === 'published';

File: api/core/src/main/java/com/mangareader/domain/manga/valueobject/ChapterStatus.java
package com.mangareader.domain.manga.valueobject;

public enum ChapterStatus {
    DRAFT, PROCESSING, SCHEDULED, PUBLISHED, HIDDEN, UNAVAILABLE, ARCHIVED;

    public boolean canTransitionTo(ChapterStatus next) {
        return switch (this) {
            case DRAFT -> next == PROCESSING || next == SCHEDULED || next == PUBLISHED || next == ARCHIVED;
            case PROCESSING -> next == DRAFT || next == PUBLISHED;
            case SCHEDULED -> next == DRAFT || next == PUBLISHED || next == ARCHIVED;
            case PUBLISHED -> next == HIDDEN || next == UNAVAILABLE || next == ARCHIVED;
            case HIDDEN, UNAVAILABLE -> next == PUBLISHED || next == ARCHIVED;
            case ARCHIVED -> next == DRAFT;
        };
    }
}
